cli/flox-rust-sdk/src/providers/publish.rs

```diff
@@ -248,38 +236,95 @@ where
     }
 }
 
+pub fn get_derivation_info(
+    path: &str,
+) -> Result<serde_json::Map<String, serde_json::Value>, PublishError> {
+    let mut info_cmd = Command::new(&*NIX_BIN);
+    info_cmd.arg("derivation").arg("show").arg(path);
+    let output = info_cmd
+        .output()
+        .map_err(|e| PublishError::MissingDerviationInfo(e.to_string()))?;
+
+    let stdout = String::from_utf8_lossy(&output.stdout);
+    let json: serde_json::Value = serde_json::from_str(&stdout)
+        .map_err(|e| PublishError::MissingDerviationInfo(e.to_string()))?;
+    Ok(json.as_object().unwrap().clone())
+}
```

Comment:
Agree with matthew that there are an excessive number of `unwrap`s and assumption of structure.
Usually if you have a known structure you can just define it as a type and derive a deserialzer for it.
Its a little bit trickier here due to the outputs being flattened in, but still worthwhile to avoid missing fields.

<details><summary>Diff for refactoring to a deserialized struct: </summary>

```diff
diff --git a/cli/flox-rust-sdk/src/providers/publish.rs b/cli/flox-rust-sdk/src/providers/publish.rs
index 2229beb30..3b9854f3d 100644
--- a/cli/flox-rust-sdk/src/providers/publish.rs
+++ b/cli/flox-rust-sdk/src/providers/publish.rs
@@ -1,11 +1,15 @@
+use std::collections::HashMap;
 use std::error;
 use std::path::PathBuf;
 use std::process::Command;
 use std::str::FromStr;
 
-use catalog_api_v1::types::{Output, SystemEnum};
+use catalog_api_v1::types::{Output, Outputs, SystemEnum};
 use chrono::{DateTime, Utc};
 use log::debug;
+use serde::de::Error;
+use serde::Deserialize;
+use serde_json::Value;
 use thiserror::Error;
 use url::Url;
 
@@ -27,9 +31,12 @@ pub enum PublishError {
     #[error("The outputs from the build do not exist: {0}")]
     NonexistentOutputs(String),
 
-    #[error("Unable to find derivation info: {0}")]
+    #[error("Unable to get derivation info: {0}")]
     MissingDerviationInfo(String),
 
+    #[error("Unable to interpret derivation info")]
+    DeserializeDerivationInfo(#[source] serde_json::Error),
+
     #[error("The environment is in an unsupported state for publishing: {0}")]
     UnsupportedEnvironmentState(String),
 
@@ -236,9 +243,20 @@ where
     }
 }
 
-pub fn get_derivation_info(
-    path: &str,
-) -> Result<serde_json::Map<String, serde_json::Value>, PublishError> {
+#[derive(Debug, Deserialize, Clone)]
+pub struct DerivationInfo {
+    name: String,
+    // could already be  SystemEnum as well
+    system: String,
+    outputs: Vec<String>,
+    version: Option<String>,
+    description: Option<String>,
+    // Remaining attributes or the drv, including the output paths,
+    // remaining envriorment variables, and/or structuredAttrs if set.
+    rest: HashMap<String, serde_json::Value>,
+}
+
+pub fn get_derivation_info(path: &str) -> Result<(String, DerivationInfo), PublishError> {
     let mut info_cmd = Command::new(&*NIX_BIN);
     info_cmd.arg("derivation").arg("show").arg(path);
     let output = info_cmd
@@ -246,54 +264,66 @@ pub fn get_derivation_info(
         .map_err(|e| PublishError::MissingDerviationInfo(e.to_string()))?;
 
     let stdout = String::from_utf8_lossy(&output.stdout);
-    let json: serde_json::Value = serde_json::from_str(&stdout)
-        .map_err(|e| PublishError::MissingDerviationInfo(e.to_string()))?;
-    Ok(json.as_object().unwrap().clone())
+    let json: Value =
+        serde_json::from_str(&stdout).map_err(PublishError::DeserializeDerivationInfo)?;
+
+    match json {
+        Value::Object(map) if map.len() == 1 => {
+            let (drv_path, drv_info_value) =
+                map.into_iter().next().expect("Proven to have a single key");
+            let drv_info = serde_json::from_value(drv_info_value)
+                .map_err(PublishError::DeserializeDerivationInfo)?;
+            Ok((drv_path, drv_info))
+        },
+        Value::Object(map) if map.len() > 1 => Err(PublishError::DeserializeDerivationInfo(
+            serde_json::Error::custom("Multiple derivations found"),
+        )),
+        _ => Err(PublishError::DeserializeDerivationInfo(
+            serde_json::Error::custom("Invalid JSON"),
+        )),
+    }
 }
 
 pub fn check_build_metadata_from_storepath(
     pkg: &str,
     storepath: &str,
 ) -> Result<CheckedBuildMetadata, PublishError> {
-    let drv_json = get_derivation_info(storepath)?;
+    let (drv_path, drv_info) = get_derivation_info(storepath)?;
+
+    let name = drv_info.name;
 
-    // Error out if there was no derivation info found
-    if drv_json.is_empty() || drv_json.keys().next().is_none() {
-        return Err(PublishError::MissingDerviationInfo(
-            "No derivation info found".to_string(),
-        ));
-    }
-    let drv_path = drv_json.keys().next().unwrap().to_string();
-    let name = drv_json[&drv_path]["name"].as_str().unwrap().to_string();
-    let system_str = drv_json[&drv_path]["system"].as_str().unwrap();
     // TODO - There is a drv_json[drv_path]["env"]["version"] in some derivations
     // that looks like the source of the version that is then later included in
     // the 'name' Once we include 'version' in the builders, we should look into
     // grabbing it from there.  Perhaps there won't ever be a 'version' where we
     // are currently looking for it?  Same for 'description'?
-    let version = drv_json[&drv_path]["version"]
-        .as_str()
-        .map(|s| s.to_string());
-    let description = drv_json[&drv_path]["description"]
-        .as_str()
-        .map(|s| s.to_string());
-
-    let system = SystemEnum::from_str(system_str).map_err(|e| {
+    let version = drv_info.version;
+    let description = drv_info.description;
+
+    let system = SystemEnum::from_str(&drv_info.system).map_err(|e| {
         PublishError::UnsupportedEnvironmentState(format!("Unable to identify system: {e}"))
     })?;
 
-    let outputs = catalog_api_v1::types::Outputs(
-        drv_json[&drv_path]["outputs"]
-            .as_object()
-            .unwrap()
-            .iter()
-            .map(|(name, value)| Output {
-                name: name.to_string(),
-                store_path: value["path"].as_str().unwrap().to_string(),
-            })
-            .collect(),
-    );
-    let outputs_to_install: Vec<String> = outputs.iter().map(|o| o.name.clone()).collect();
+    let outputs = {
+        let mut outputs = Vec::new();
+        for output_name in drv_info.outputs.iter() {
+            let store_path_value = drv_info.rest.get(output_name).ok_or_else(|| {
+                PublishError::MissingDerviationInfo(format!(
+                    "Output {output_name} not found in derivation"
+                ))
+            })?;
+            let store_path = serde_json::from_value(store_path_value.clone())
+                .map_err(PublishError::DeserializeDerivationInfo)?;
+            let output = Output {
+                name: output_name.to_string(),
+                store_path,
+            };
+            outputs.push(output);
+        }
+        Outputs(outputs)
+    };
+
+    let outputs_to_install: Vec<String> = drv_info.outputs.clone();
 
     Ok(CheckedBuildMetadata {
         drv_path,

```

</details>

Reply:
Besides i think that the metadata should come from the lockfile/manifest, rather than stuffing it all in the derivation.
That way we might be able to avoid having three different variations of the same'ish metadata, with `DerivationInfo/Value`, `CheckedBuildMetadata`, `UserDerivationInfo`.

Reply:
Refactored as suggested and removed sourcing description and version from the derivation.  We'll get that from the manifest once it's added.